### PR TITLE
Server-wide and Channel-wide Command Disabling

### LINKED ISSUE #765: Feature: Disabling Commands Server-wide
Have the ability to enable and disable certain commands server-wide.

Current thoughts:
- Command Name: `command` (aliases: `cmd`)
- Arguments: `<command name> [enable | on | disable | off]` (if given the command name it will let the user know if it's enabled or disabled (does not require claim))
- When enabling or disabling commands (i.e., not `!cmd <command name>`) the `manage_commands` claim is required.
- Only applies to a command or a group of commands. You cannot disable a sub-command, i.e., `!tag add`.
- Possible idea: have another claim to bypass disabled commands? Probably named `bypass_disabled_commands`?
- Command Service
  - Might have to rename from `CommandMetricsService` just to `CommandService` if we don't want to add another service.
  - On each command run, the bot will request from the API whether or not that command is disabled. If so, the user will be notified that it's been disabled server-wide.
- Additions to Commands Route
  - **[GET]** `bot/commands/{GuildId}/{CommandName}` - Will check against a `List<string>` from the API and will return a boolean, `true` if enabled, `false` if disabled.

### PR DESCRIPTION
**Pull request by jkriste**
# Command Restrictions
This PR adds a new cog and a way to disable commands in specific channels or server-wide. Resolves #765.

## Features
- Command Cog
  - **Name**: `command` (aliases: `cmd`)
    - Checks the status of a command for a guild and channel.
    - **Arguments**: `<command name>`
    - **Examples**: `!command help`, `!command slots`
  - **Sub-commands**:
    - **Name**: `enable` (aliases: `on`)
      - Enables this command in a specific channel or server-wide.
      - **Arguments**: `<command name> [channel]`
      - **Required Claims**: `manage_commands`
      - **Examples**: `!command enable search #general-chat`, `!command enable slots`
    - **Name**: `disable` (aliases: `off`)
      - Disables this command in a specific channel or server-wide. 
      - **Arguments**: `<command name> [channel] [silently fail]`
      - **Required Claim**: `manage_commands`
      - If silently fail is `true`, the user will not be notified that the command is disabled.
      - **Examples**: `!command disable trivia true`, `!command disable slots #clem-bot-discussion`
- Two new claims:
  - `manage_commands`: Allows for enabling and disabling of commands.
  - `bypass_disabled_commands`: Allows for a user or role to bypass **ANY** disabled commands.
- Cogs can optionally opt-out of allowing users to disable the command.
  - **Example**:
  ```py
  @ext.command()
  @ext.allow_disable(False)  # by default, this is set to True and is not necessary to allow disabling
  async def my_command(self, ctx: ext.ClemBotCtx) -> None:
      # ...
  ```
## API Routes
- **[GET]** `bot/commands/status`
  - Checks whether the given command is disabled in either the given guild or channel.
  - **Body**:
  ```cs
  command_name: str
  guild_id: int
  channel_id: int
  ```
  - **Returns**:
  ```cs
  disabled: bool
  silently_fail: bool
  ```
- **[GET]** `bot/commands/details`
  - Gets the details for command restrictions for the given command name in the given guild and channel.
  - **Body**:
  ```cs
  command_name: str
  guild_id: int
  channel_id: int
  ```
  - **Returns**:
  ```cs
  command_name: str
  disabled: bool
  guild_id: int
  channel_ids: list[int]
  ```
- **[PUT]** `bot/commands/disable`
  - Disables the given command. 
  - If the given command was previously disabled 1 or more channels and is now being disabled server-wide, the record for the channel disable will be removed.
  - **Body**:
  ```cs
  command_name: str
  guild_id: int
  channel_id: int | None
  silently_fail: bool = False
  ```
  - **Returns**: `None`
- **[DELETE]** `bot/commands/enable`
  - Enables the given command.
  - If the given command was previously disabled 1 or more channels and is being enabled server-wide, the record for the channel disable will be removed.
  - **Body**:
  ```cs
  command_name: str
  guild_id: int
  channel_id: int | None
  ```
  - **Returns**: `None`
## Other
- Disallowed disabling for the `!help` and `!command` commands.
- Fixed a small typo in the `tag_cog.py` documentation (it was really bothering me).
- Renamed `CommandMetricsService` to `CommandService`.
## To-Do
- [x] Create a migration to include the new claims and table.
- [x] Finish enabling and disabling in the API.
- [x] Finish enabling and disabling in the bot.
- [x] Move claims checking to a service (as requested by @Jay-Madden).
- [x] Check for `bypass_disabled_commands` claim pre-status.
- [x] Test
- [x] Add documentation to Wiki.
- [x] ~Beg @Jay-Madden to review the PR several times.~

### CODE REVIEW

File: ClemBot.Bot/bot/cogs/command_cog.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        embed.add_field(name="Name", value=f"`{cmd.name}`")
+        embed.add_field(name="Allows Disabling", value=cmd.allow_disable, inline=False)
+        embed.add_field(name=name, value=value)
+        if cmd.allow_disable:
+            opp_mode = "enable" if model.disabled else "disable"
+            embed.set_footer(
+                text=f"You can {opp_mode} this command by typing "
+                f'"{ctx.prefix}cmd {opp_mode} {cmd.name}"'
+            )
+        await ctx.send(embed=embed)
```
Comment (jkriste) [inline ClemBot.Bot/bot/cogs/command_cog.py:47]: Perhaps make this embed deletable?

File: ClemBot.Bot/bot/api/commands_route.py [hunk 2]
```diff
@@ -20,3 +21,54 @@ async def add_command_invocation(
         }
 
         await self._client.post("bot/commands", data=json, **kwargs)
+
+    async def get_status(
+        self, guild_id: int, channel_id: int, command_name: str, **kwargs: t.Any
+    ) -> tuple[bool, bool]:
```
Comment (Jay-Madden) [inline ClemBot.Bot/bot/api/commands_route.py:27]: This function should return a named object/class.

**Review by Jay-Madden — CHANGES_REQUESTED** [review]
Great PR, some organizational stuff up front, also add some new lines in the python code in some places you deem to be best. its very dense

**Review by Iapetus-11 — CHANGES_REQUESTED** [review]
Please space out your Python code to make it more readable (you actually do in some places, just not everywhere :madge_distort:). My general rule of thumb is newline after dedentation or before returns.

File: ClemBot.Bot/bot/errors.py [hunk 3]
```diff
@@ -49,6 +49,21 @@ def __init__(self, message: str):
...
+
+
+class SilentCommandRestrictionError(CommandError):
+    """
+    Raised if a user attempts to use a command that is disabled silently
+    """
+
+
 class ConversionError(CommandError):
     def __init__(self, message: typing.Optional[str] = None):
```
[the remark this answers is not in the document]
Comment (jkriste) [reply ClemBot.Bot/bot/errors.py:68]: I did not write the ConversionError class 😆

**Review by Jay-Madden — APPROVED** [review]
lgtm

**Pull request merged**